rust_icu_unum/src/lib.rs

```diff
@@ -20,10 +20,68 @@ use {
     rust_icu_common as common, rust_icu_sys as sys,
     rust_icu_sys::versioned_function,
     rust_icu_sys::*,
-    rust_icu_ustring as ustring,
-    std::{convert::TryFrom, os::raw, ffi, ptr},
+    rust_icu_uloc as uloc, rust_icu_ustring as ustring,
+    std::{convert::TryFrom, ffi, os::raw, ptr},
 };
 
-#[cfg(test)]
-mod tests {
+#[derive(Debug)]
+pub struct UNumberFormat {
+    rep: ptr::NonNull<sys::UNumberFormat>,
+}
+
+impl Drop for UNumberFormat {
+    /// Implements `unum_close`
+    fn drop(&mut self) {
+        unsafe { versioned_function!(unum_close)(self.rep.as_ptr()) };
+    }
 }
+
+impl UNumberFormat {
+    /// Implements `unum_open`
+    pub fn try_new_ustring(
+        style: sys::UNumberFormatStyle,
+        pattern: ustring::UChar,
+        locale: &uloc::ULoc,
+    ) -> Result<UNumberFormat, common::Error> {
+        let mut status = common::Error::OK_CODE;
+        let mut parse = common::NO_PARSE_ERROR.clone();
+        let loc = locale.as_c_str();
+
+        // Unsafety note: all variables should be valid.
+        let rep = unsafe {
+            assert!(common::Error::is_ok(status));
+            assert!(common::parse_ok(parse).is_ok());
```

Comment:
Out of curiosity, is this really needed? It does have some (minuscule) cost, and I can't think of any circumstances in which the assertion could ever fail.

Reply:
If I didn't have an assert here, I'd have a comment that says essentially the same thing.  So, why not have an assert to begin with? it's the same sort of precondition commitment, but with more teeth.

And every compiler worth its salt should be able to be told to not generate assertion code if that's needed.